producer/writer/consumer_service_writer.go

```diff
@@ -0,0 +1,311 @@
+
+	cs             topic.ConsumerService
+	numberOfShards uint32
+	ps             placement.Service
+	shardWriters   []shardWriter
+	opts           Options
+	logger         log.Logger
+
+	dataFilter      producer.FilterFunc
+	router          ackRouter
+	consumerWriters map[string]consumerWriter
+	isClosed        bool
+	doneCh          chan struct{}
+	wg              sync.WaitGroup
+	m               consumerServiceWriterMetrics
+
+	processPlacementFn processPlacementFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerServiceWriter(
+	cs topic.ConsumerService,
+	numberOfShards uint32,
+	mPool messagePool,
+	opts Options,
+) (consumerServiceWriter, error) {
+	ps, err := opts.ServiceDiscovery().PlacementService(cs.ServiceID(), nil)
+	if err != nil {
+		return nil, err
+	}
+	router := newAckRouter(int(numberOfShards))
+	w := &consumerServiceWriterImpl{
+		cs:              cs,
+		numberOfShards:  numberOfShards,
+		ps:              ps,
+		shardWriters:    initShardWriters(router, cs, numberOfShards, mPool, opts),
+		opts:            opts,
+		logger:          opts.InstrumentOptions().Logger(),
+		dataFilter:      acceptAllFilter,
+		router:          router,
+		consumerWriters: make(map[string]consumerWriter),
+		isClosed:        false,
+		doneCh:          make(chan struct{}),
+		m:               newConsumerServiceWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	w.processPlacementFn = w.processPlacement
+	return w, nil
+}
+
+func initShardWriters(
+	router ackRouter,
+	cs topic.ConsumerService,
+	numberOfShards uint32,
+	mPool messagePool,
+	opts Options,
+) []shardWriter {
+	sws := make([]shardWriter, numberOfShards)
+	for i := range sws {
+		switch cs.ConsumptionType() {
+		case topic.Shared:
+			sws[i] = newSharedShardWriter(uint32(i), router, mPool, opts)
+		case topic.Replicated:
+			sws[i] = newReplicatedShardWriter(uint32(i), numberOfShards, router, mPool, opts)
+		}
+	}
+	return sws
+}
+
+func (w *consumerServiceWriterImpl) Write(d producer.RefCountedData) error {
+	shard := d.Shard()
+	if shard >= w.numberOfShards {
+		w.m.invalidShard.Inc(1)
+		return fmt.Errorf("could not write data for shard %d which is larger than max shard id %d", shard, w.numberOfShards)
+	}
+	if d.Filter(w.dataFilter) {
+		w.shardWriters[shard].Write(d)
+	}
+	// It is not an error if the data does not pass the filter.
+	return nil
+}
+
+func (w *consumerServiceWriterImpl) Init() error {
+	var (
+		pw      placement.Watch
+		err     error
+		retrier = retry.NewRetrier(w.opts.PlacementWatchRetryOptions().SetForever(true))
+	)
+	retrier.Attempt(func() error {
+		pw, err = w.ps.Watch()
+		if err != nil {
+			// Unexpected, but if it ever happens, we will retry forever to recreate the watch.
+			w.logger.Errorf(
+				"could not create placement watch for %s, retry later: %v",
+				w.cs.ServiceID().String(),
+				err,
+			)
+			return err
+		}
+		return nil
+	})
+	err = w.waitForInitialPlacement(pw)
```

Comment:
Use `https://github.com/m3db/m3cluster/blob/master/kv/util/runtime/value.go` for watching placement?

Reply:
Not sure if I follow. I was suggesting using the `Value` interface to replace the logic of getting initial value and watching it here since that's already taken care for you by the runtime value implementation. It is not tied to placement in any way. It's basically a simple interface for watching arbitrary runtime value updates and provides options for configuring unmarshaling functions etc.
```
// Value is a value that can be updated during runtime.
type Value interface {
	// Key is the key associated with value.
	Key() string

	// Watch starts watching for value updates.
	Watch() error

	// Unwatch stops watching for value updates.
	Unwatch()
}
```

Reply:
I see, I feel that's an overkill here as I probably will never call Unwatch, directly putting the init logic here seems more intuitive.

Reply:
The runtime value is not just for initialization though, I believe you can remove all the logic in `Init`, `waitForInitialPlacement`, `waitPlacementUpdatesForever`, and `processPlacementWatchUpdate` if you use it, which seems like a pretty good win.

It seems you are listening on `doneCh` in `watchPlacementUpdateForever` and return if `doneCh` is closed, which is basically the equivalent of `UnWatch`.

Reply:
Yeah seems there is some value in reusing the `Value` interface, although the `Watch` in the `Value` interface does not seem to handle retry forever on watch creation errors, but that is something that could be added.
One thing about the `Value` interface is that it uses `kv.Store`, but I would prefer to use `PlacementService` here as I don't want to rely on the knowledge of how `PlacementService` internally configs `kv.Store`, in another similar case I would like to use `TopicService` as well, I think we should maybe use more options to not depend on `kv.Store` internally.

Reply:
Hm do you really need to retry forever when creating watches? IOW, watch creation never fails unless you are creating a watch on a closed store or sth, in which case retry is not going to help no?

Can we replace `kv.Store` in `Value` implementation with a watchable interface that both `kv.Store` and `PlacementService` implements?